openmdao/core/driver.py

```diff
@@ -310,16 +314,66 @@ def _setup_driver(self, problem):
                     sz = sizes[owner, i]
 
                 if vname in dv_set:
-                    dv_dict[vname] = (owner, sz)
+                    remote_dv_dict[vname] = (owner, sz)
+
+                # Note that design vars are not distributed.
                 elif distributed:
                     idx = model._var_allprocs_abs2idx['nonlinear'][vname]
                     dist_sizes = model._var_sizes['nonlinear']['output'][:, idx]
-                    dist_con_dict[vname] = (idx, dist_sizes)
+
+                    # Determine which indices are on our proc.
+                    rank = model.comm.rank
+                    size = dist_sizes.size
+                    offsets = np.cumsum(dist_sizes)
+
+                    if vname in self._responses:
+                        resp_dict = self._responses[vname]
+                        indices = resp_dict['indices']
+
+                        if indices is not None:
+                            local_indices = []
+                            true_sizes = np.zeros(size, dtype=INT_DTYPE)
+                            for index in indices:
+                                if index < 0:
+                                    # Support for negative indices. Convert to positive index.
+                                    index = index + np.sum(dist_sizes)
+                                irank = np.argwhere(offsets > index)[0][0]
+                                true_sizes[irank] += 1
+                                if rank == irank:
+                                    new_index = index - offsets[irank] + dist_sizes[irank]
+                                    local_indices.append(new_index)
+
+                            indices = local_indices
+                            dist_sizes = true_sizes
+
+                        dist_con_dict[vname] = (indices, dist_sizes)
+
+                    if vname in self._objs:
```

Comment:
Seems like this whole block could go away since the code is identical to the block above, and you could just check inside of the `if vname in self._responses` block to see if vname is in self._objs and if so set the dist_obj_dict entry there (since self._responses contains both the constraints and the objectives).

Reply:
I was thinking we allowed a con and obj to be declared on the same var, but apparently we don't. I should just combine these into one dict for responses then.

Reply:
I thought we could to that too.